File: cpp/longest-palindromic-substring.cpp
// quadratic runtime, constant space
class Solution {
public:
    string longestPalindrome(string s) {
        int bestLength = 0;
        string bestString = "";

        const int size = s.length();

        // middle-out aglorithm lol
        for (int mid = 0; mid < size; ++mid) {
            for (int x = 0; mid - x >= 0 and mid + x < size; ++x) { 
                if (s[mid - x] != s[x + mid]) {
                    break; // not a palindrome
                }
                // if we didn't break, then we have a palindrome which spans
                // from the mid character plus two character substrings of 
                // length x to the left and right of the mid character.
                int length = 2 * x + 1;
                if (length > bestLength) {
                    bestLength = length;
                    bestString = s.substr(mid - x, length);
                }
            }
        }

        // odd length string case
        for (int mid = 0; mid < size - 1; ++mid) {
            for (int x = 1; mid - x + 1 >= 0 and mid + x < size; ++x) {
                if (s[mid - x + 1] != s[x + mid]) {
                    break; // not a palindrome
                }
                int length = 2 * x;
                if (length > bestLength) {
                    bestLength = length;
                    bestString = s.substr(mid - x + 1, length);
                }
            }
        }

        return bestString;
    }
};

File: py/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        size = len(s)
        if size < 2:
            return s

        bestString = ''
        bestLength = 0

        for i in range(size):
            # odd case
            checkString = self.getPalindrome(s, i, i, size)
            checkLength = len(checkString)
            if checkLength > bestLength:
                bestString = checkString
                bestLength = checkLength

            # even case
            checkString = self.getPalindrome(s, i, i + 1, size)
            checkLength = len(checkString)
            if checkLength > bestLength:
                bestString = checkString
                bestLength = checkLength

        return bestString

    def getPalindrome(self, s: str, left, right, size: int) -> str:
        while left >= 0 and right < size and s[left] == s[right]:
            left -= 1 # move left
            right += 1 # move right
        return s[left + 1 : right]


print(Solution().longestPalindrome("cbbd")) # bb
print(Solution().longestPalindrome("ccd")) # cc
print(Solution().longestPalindrome("bacaba")) # bacab
print(Solution().longestPalindrome("abacaba")) # abacaba
print(Solution().longestPalindrome("a")) # a
print(Solution().longestPalindrome("ccc")) # ccc
print(Solution().longestPalindrome("xx")) # xx
print(Solution().longestPalindrome("abacab")) # bacab